tools/pip-compile-wrapper.py

```diff
@@ -0,0 +1,300 @@
+        target_dir: Path = source_path.parent
+        with source_path.open("rt") as target_file:
+            for line in target_file:
+                line = line.strip()
+                if line.startswith("-r"):
+                    required = (
+                        target_dir.joinpath(line.lstrip("-r").strip())
+                        .resolve()
+                        .relative_to(REQUIREMENTS_DIR)
+                        .stem
+                    )
+                    SOURCE_DEPENDENCIES[source_name].add(required)
+
+
+_resolve_source_dependencies()
+
+
+def _run_pip_compile(
+    source_name: str,
+    upgrade_all: bool = False,
+    upgrade_packages: Optional[Set[str]] = None,
+    verbose: bool = False,
+    dry_run: bool = False,
+) -> None:
+    assert_msg = "Only one of `upgrade_all` or `upgrade_packages` may be given."
+    assert not (upgrade_all and upgrade_packages), assert_msg
+
+    pip_compile_exe = which("pip-compile")
+    if not pip_compile_exe:
+        raise RuntimeError("pip-compile missing. This shouldn't happen.")
+
+    if not upgrade_packages:
+        upgrade_packages = set()
+
+    source_path = SOURCE_PATHS[source_name]
+    target_path = TARGET_PATHS[source_name]
+
+    upgrade_packages_cmd: List[str] = []
+    if upgrade_packages:
+        packages_in_target = {
+            package_name
+            for package_name, _ in _get_requirement_packages(source_name, TargetType.ALL)
+        }
+
+        upgrade_packages_cmd = list(
+            chain.from_iterable(
+                zip(repeat("--upgrade-package"), upgrade_packages.intersection(packages_in_target))
+            )
+        )
+    upgrade_all_cmd: List[str] = []
+    if upgrade_all:
+        upgrade_all_cmd = ["--upgrade"]
+
+    dry_run_cmd = ["--dry-run"] if dry_run else []
+
+    command = [
+        pip_compile_exe,
+        "--verbose" if verbose else "--quiet",
+        *dry_run_cmd,
+        "--no-index",
+        *upgrade_packages_cmd,
+        *upgrade_all_cmd,
+        "--output-file",
+        str(target_path),
+        str(source_path),
+    ]
+    print(f"Compiling {source_path.name}...", end="")
+    sys.stdout.flush()
+    if verbose:
+        print(f"\nRunning command: {' '.join(shlex.quote(c) for c in command)}")
+    env = os.environ.copy()
+    env["CUSTOM_COMPILE_COMMAND"] = "'./deps compile' (for details see README)"
+    process = subprocess.run(
+        command, capture_output=(not verbose), cwd=str(source_path.parent), env=env
+    )
+    if process.returncode == 0:
+        print("\b\b Success.")
+        return
+    print("\b\b Error!")
+    if not verbose:
+        print(process.stdout.decode())
+        print(process.stderr.decode())
+    process.check_returncode()
+
+
+def _resolve_deps(source_names: Iterable[str]) -> List[Set[str]]:
+    requirements = {
+        source: dependencies.intersection(source_names)
+        for source, dependencies in SOURCE_DEPENDENCIES.items()
+        if source in source_names
```

Comment:
what if `source_names` has an unknown dependency? Shouldn't the following check be used instead?

```python
unknown = set(source_names) - set(SOURCE_DEPENDENCIES)
assert not unknown, 'Unknown dependencies {unknown}'
```

(With the above I'm assuming that the `SOURCE_DEPENDENCIES` always has an entry, even if it's an empty list/set, which seems the case since there is a `Dummy access to create the key`)

Reply:
btw, shouldn't `source_names` have a name like `requirement_files`?

Reply:
I guess that check wouldn't hurt, but since it's only intended to be used internally 

It's `_names` since that is only the name part without extension.
I picked `source` and `target` to make it more obvious which is which (since both are technically 'requriements').